File: app/src/main/java/com/example/sasham/testproject/themes/ThemesListingPresenterImp.java
package com.example.sasham.testproject.themes;

import com.example.sasham.testproject.model.Theme;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.schedulers.Schedulers;

public class ThemesListingPresenterImp implements ThemesListingPresenter {

    private ThemesListingView view;
    private ThemesListingInteractor themesListingInteractor;
    private List<Theme> loadedThemes = new ArrayList<>();
    private CompositeDisposable compositeDisposable=new CompositeDisposable();
    private int currentPage;


    @Inject
    public ThemesListingPresenterImp(ThemesListingInteractor themesListingInteractor) {
        this.themesListingInteractor = themesListingInteractor;
    }

    @Override
    public void setView(ThemesListingView view) {
        this.view = view;
    }

    @Override
    public void firstPage() {
        currentPage = 1;
        loadedThemes.clear();
        loadThemes();
    }

    @Override
    public void nextPage() {
        currentPage++;
        loadThemes();
    }

    @Override
    public void destroy() {
        view = null;
        compositeDisposable.clear();
    }

    @Override
    public void loadNewData() {
        firstPage();
    }


    private void loadThemes() {
        view.onLoading();
        Disposable disposable = themesListingInteractor.fetchThemes(currentPage)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Consumer<List<Theme>>() {
                               @Override
                               public void accept(List<Theme> themes) throws Exception {
                                   loadedThemes.addAll(themes);
                                   displayThemes();
                               }
                           },
                        new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) throws Exception {
                                view.onError(throwable.getMessage());
                            }
                        });

//        compositeDisposable.add(disposable);
    }

    private void displayThemes() {
        if (isViewAttached()) {
            view.showThemes(loadedThemes);
        }
    }

    private boolean isViewAttached() {
        return view != null;
    }

}

File: app/src/main/java/com/example/sasham/testproject/themes/ThemesListingPresenterImp.kt
package com.example.sasham.testproject.themes

import com.arellomobile.mvp.InjectViewState
import com.arellomobile.mvp.MvpPresenter
import com.example.sasham.testproject.App
import com.example.sasham.testproject.model.DataRepository
import com.example.sasham.testproject.model.Theme
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import java.util.*
import javax.inject.Inject

@InjectViewState
class ThemesListingPresenterImp : MvpPresenter<ThemesListingView>(), ThemesListingPresenter {

    @Inject
    lateinit var data: DataRepository

    private val loadedThemes = ArrayList<Theme>()
    private val compositeDisposable = CompositeDisposable()
    private var currentPage: Int = 0

    init {
        App.instance!!.appComp!!.inject(this)
    }

    override fun firstPage() {
        currentPage = 1
        loadedThemes.clear()
        loadThemes()
    }

    override fun nextPage() {
        currentPage++
        loadThemes()
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }

    override fun loadNewData() {
        firstPage()
    }


    private fun loadThemes() {

        val disposable = data.themes(currentPage.toString())
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ themes ->
                    loadedThemes.addAll(themes)
                    viewState.showThemes(loadedThemes)
                },
                        { throwable -> viewState.message(throwable.message) })

        compositeDisposable.add(disposable);
    }
}
